mewarpx/mewarpx/emission.py

```diff
@@ -1197,6 +1197,100 @@ def get_normals(self, x, y, z):
         return normals
 
 
+class ZPlanePatchyEmitter(object):
+    """Injection for a patchy cathode."""
+
+    geoms = ['XZ']
+
+    def __init__(self, patchy_cathode, transverse_fac=1.0, **kwargs):
+        """Initialize an emitter for a planar patchy cathode.
+
+        Arguments:
+            patchy_cathode (:class:`mewarpx.assemblies.PatchyCathode`): Patchy
+                cathode object, used to obtain work functions, patch size and z
+                coordinate/direction.
+            transverse_fac (float): Scale the transverse energy distribution by
+                this factor. Default 1. See
+                :func:`mewarpx.utils_store.util.get_velocities` for details.
+            kwargs (dict): Any other keyword arguments supported by the
+                Emitter constructor (such as "emission_type").
+        """
+        self.patchy_cathode = patchy_cathode
+
+        # The tricky part here is that we need to calculate the appropriate
+        # size for each patch set knowing that the domain is not necessarily
+        # an even integer multiple of the patch size. Note that the
+        # PatchyCathode is hard coded to have a high WF patch start at x=0.
+        patch_num = abs(mwxrun.xmax) / self.patchy_cathode.patch_size
+        full_patches = np.floor(patch_num)
+        if full_patches % 2.0:
+            high_wf_patches = (full_patches + 1.0) / 2.0
+        else:
+            high_wf_patches = full_patches / 2.0 + (patch_num - full_patches)
+
+        patch_num = abs(mwxrun.xmin) / self.patchy_cathode.patch_size
+        full_patches = np.floor(patch_num)
+        if full_patches % 2.0:
+            high_wf_patches += (full_patches - 1.0) / 2.0 + (patch_num - full_patches)
+        else:
+            high_wf_patches += full_patches / 2.0
+
+        high_wf_length = high_wf_patches * self.patchy_cathode.patch_size
+        low_wf_length = mwxrun.xmax - mwxrun.xmin - high_wf_length
+
+        # Default initialization for the two sets of patches
+        self.high_wf = ZPlaneEmitter(
+            conductor=self.patchy_cathode.cathode_list[0],
+            xmin=-high_wf_length/2.0, xmax=high_wf_length/2.0, **kwargs
+        )
+        self.low_wf = ZPlaneEmitter(
+            conductor=self.patchy_cathode.cathode_list[1],
+            xmin=-low_wf_length/2.0, xmax=low_wf_length/2.0, **kwargs
+        )
+
+        # Finally, we need to overwrite the ``_get_xv_coords()`` function
+        # for the two emitters in order to properly place the emitted electrons
+        # in front of patches
+        self.high_wf_orig_coord_func = self.high_wf._get_xv_coords
+        self.high_wf._get_xv_coords = self.move_particles_high_wf_patch
+        self.low_wf_orig_coord_func = self.low_wf._get_xv_coords
+        self.low_wf._get_xv_coords = self.move_particles_low_wf_patch
+
+    def move_particles_high_wf_patch(self, npart, m, rseed):
```

Comment:
Since this method and the corresponding `move_particles_low_wf_patch` are not actually used by the parent `ZPlanePatchyEmitter` class, would it make more sense semantically to create a new class for the patches that inherits from `ZPlaneEmitter` and overwrites / redefines `_get_xv_coords`?